presto-main/src/main/java/com/facebook/presto/operator/aggregation/AggregationCompiler.java

```diff
@@ -254,4 +301,138 @@ private static Method getCombineFunction(Class<?> clazz, Class<?> stateClass)
+            List<Type> inputTypes = boundSignature.getArgumentTypes().stream().map(x -> typeManager.getType(x)).collect(toImmutableList());
+            Type outputType = typeManager.getType(boundSignature.getReturnType());
+
+            AggregationFunction aggregationAnnotation = definitionClass.getAnnotation(AggregationFunction.class);
+            requireNonNull(aggregationAnnotation, "aggregationAnnotation is null");
+
+            DynamicClassLoader classLoader = new DynamicClassLoader(definitionClass.getClassLoader());
+
+            AggregationMetadata metadata;
+            AccumulatorStateSerializer<?> stateSerializer = new StateCompiler().generateStateSerializer(stateClass, classLoader);
+            Type intermediateType = stateSerializer.getSerializedType();
+            Method intermediateInputFunction = getIntermediateInputFunction(definitionClass, stateClass);
+            Method combineFunction = getCombineFunction(definitionClass, stateClass);
+            AccumulatorStateFactory<?> stateFactory = new StateCompiler().generateStateFactory(stateClass, classLoader);
+
+            try {
+                MethodHandle inputHandle = lookup().unreflect(inputFunction);
+                MethodHandle intermediateInputHandle = intermediateInputFunction == null ? null : lookup().unreflect(intermediateInputFunction);
+                MethodHandle combineHandle = combineFunction == null ? null : lookup().unreflect(combineFunction);
+                MethodHandle outputHandle = outputFunction == null ? null : lookup().unreflect(outputFunction);
+                metadata = new AggregationMetadata(
+                        generateAggregationName(signature.getName(), outputType.getTypeSignature(), signaturesFromTypes(inputTypes)),
+                        getParameterMetadata(inputFunction, inputTypes),
+                        inputHandle,
+                        getParameterMetadata(intermediateInputFunction, inputTypes),
+                        intermediateInputHandle,
+                        combineHandle,
+                        outputHandle,
+                        stateClass,
+                        stateSerializer,
+                        stateFactory,
+                        outputType,
+                        aggregationAnnotation.approximate());
+            }
+            catch (IllegalAccessException e) {
+                throw Throwables.propagate(e);
+            }
+
+            AccumulatorFactoryBinder factory = new LazyAccumulatorFactoryBinder(metadata, classLoader);
+
+            return new InternalAggregationFunction(signature.getName(),
+                    inputTypes,
+                    intermediateType,
+                    outputType,
+                    decomposable,
+                    approximate,
+                    factory);
+        }
+
+        public InternalAggregationFunction getOnlySpecialization()
+        {
+            return getOnlySpecialization(new TypeRegistry());
```

Comment:
65 occurences of "new TypeRegistry" in our code. It doesn't look like singleton to me ;-)
I agree with the idea but current codebase is not ready to avoid this in simple way.